search-app/src/cmr/search/services/virtual_directory.clj

```diff
@@ -0,0 +1,225 @@
+  [level]
+  (format "%s%s%s"
+          (if (:year level) (str (:year level) "/") "")
+          (if (:month level) (str (:month level) "/") "")
+          (if (:day level) (str (:day level) "/") "")))
+
+(defn- remove-trailing-slash
+  "Removes trailing slash from a string."
+  [str]
+  (if (string/ends-with? str "/")
+    (subs str 0 (dec (.length str)))
+    str))
+
+(defn build-link
+  "Builds a link to navigate to the next level in the virtual directory structure."
+  [context concept-id value current-level]
+  (remove-trailing-slash (format "%s/%s/%s%s" (base-url context) concept-id
+                                 (level->str current-level) value)))
+
+(defn build-remove-link
+  "Builds a remove link."
+  [context concept-id current-level]
+  (remove-trailing-slash (format "%s/%s/%s" (base-url context) concept-id
+                                 (level->str current-level))))
+
+(defn build-group-node
+  "Builds a group node."
+  [title children]
+  {:title title
+   :type :group
+   :children children})
+
+(defn build-title-type-links-to-remove
+  "Build the title type and links to remove for the parent node."
+  [context concept-id value children link-level]
+  {:title value
+   :type :filter
+   :links {:remove (build-remove-link context concept-id link-level)}
+   :children children})
+
+(defn build-group-and-filter-node
+  "Builds a group node with a filter node as a child."
+  [context concept-id title parent-value children link-level]
+  (let [group-node [(build-group-node title children)]
+        parent-value-node (build-title-type-links-to-remove context concept-id parent-value
+                                                            group-node link-level)]
+    parent-value-node))
+
+(defn time-ranges->clj-time-interval
+  "Returns the next interval based on the values in the map."
+  [time-ranges]
+  (let [all-keys (set (keys time-ranges))]
+    (if-not (contains? all-keys :year)
+      nil
+      (if-not (contains? all-keys :month)
+        (clj-time/years 1)
+        (if-not (contains? all-keys :day)
+          (clj-time/months 1)
+          (if-not (contains? all-keys :hour)
+            (clj-time/days 1)
+            (clj-time/hours 1)))))))
+
+(defn time-ranges->next-interval
+  "Returns the next interval based on the values in the map."
+  [time-ranges]
+  (let [all-keys (set (keys time-ranges))]
+    (if-not (contains? all-keys :year)
+      :year
+      (if-not (contains? all-keys :month)
+        :month
+        (if-not (contains? all-keys :day)
+          :day
+          (when-not (contains? all-keys :hour)
+            :hour))))))
+
+(def num-urls
+  "Number of download URLs to return"
+  100)
+
+(defn- build-top-level-response
+  "Builds the final JSON response."
+  [context concept-id {:keys [year month day hour] :as time-ranges} children download-urls]
+  (let [interval-granularity (time-ranges->next-interval time-ranges)
+        hour-node (if (= :hour interval-granularity)
```

Comment:
I'd split all of these (`hour-node`, `day-node`, etc.) into their own private functions and just call them here to get the assignments. Too much nested code in the `let`, etc.

Reply:
I wrote the huge prototype disclaimer to allow horrible ugly code... But yes this whole function needs to be rewritten and there's probably a lot of commonality that can be pulled out.